src/XMakeBuildEngine/Definition/Toolset.cs

```diff
@@ -354,33 +347,34 @@ private Toolset(INodePacketTranslator translator)
         }
 
         /// <summary>
-        /// Returns a copy of the list of search paths for a MSBuildExtensionsPath* property.
-        /// <returns>List of search paths, else null</returns>
+        /// Returns a copy of the list of search paths for a given property.
+        /// <returns>List of search paths or ProjectImportPathMatch.None if empty</returns>
         /// </summary>
-        internal List<string> GetMSBuildExtensionsPathSearchPathsFor(string refKind)
+        internal ProjectImportPathMatch GetProjectImportSearchPaths(string expression)
         {
-            if (string.IsNullOrEmpty(refKind))
-                return null;
+            if (string.IsNullOrEmpty(expression) || ImportPropertySearchPathsTable == null)
+            {
+                return ProjectImportPathMatch.None;
+            }
 
-            List<string> paths;
-            if (MSBuildExtensionsPathSearchPathsTable != null && MSBuildExtensionsPathSearchPathsTable.TryGetValue(refKind, out paths))
+            foreach (var searchPath in _propertySearchPathsTable)
             {
-                return new List<string>(paths);
+                if (expression.IndexOf($"$({searchPath.Key})", StringComparison.OrdinalIgnoreCase) >= 0)
```

Comment:
This would work fine. It would first try as is and then try with $(FallbackProperty) replaced with something else. $(NonFallbackProperty) just behaves normally.